support/mesh/mesh_channel_core/src/mpsc.rs

```diff
@@ -0,0 +1,979 @@
+            queue: ReceiverQueue(queue),
+            ports: PortHandlerList::new(),
+            terminated: false,
+        }
+    }
+}
+
+fn trace_channel_error(err: &ChannelError) {
+    tracing::error!(
+        error = err as &dyn std::error::Error,
+        "channel closed due to error"
+    );
+}
+
+impl<T> futures_core::Stream for Receiver<T> {
+    type Item = T;
+
+    fn poll_next(self: std::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Poll::Ready(match std::task::ready!(self.get_mut().poll_recv(cx)) {
+            Ok(t) => Some(t),
+            Err(RecvError::Closed) => None,
+            Err(RecvError::Error(err)) => {
+                trace_channel_error(&err);
+                None
+            }
+        })
+    }
+}
+
+impl<T> futures_core::FusedStream for Receiver<T> {
+    fn is_terminated(&self) -> bool {
+        self.0.terminated
+    }
+}
+
+#[derive(Debug)]
+struct PortHandlerList(Vec<PortWithHandler<RemotePortHandler>>);
+
+impl PortHandlerList {
+    fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    fn remove_closed(&mut self) -> Result<(), ChannelError> {
+        let mut r = Ok(());
+        self.0.retain(|port| match port.is_closed() {
+            Ok(true) => false,
+            Ok(false) => true,
+            Err(err) => {
+                let err = ChannelError::from(err);
+                if r.is_ok() {
+                    r = Err(err);
+                } else {
+                    trace_channel_error(&err);
+                }
+                false
+            }
+        });
+        r
+    }
+
+    fn into_ports(self) -> Vec<Port> {
+        self.0
+            .into_iter()
+            .map(|port| port.remove_handler().0)
+            .collect()
+    }
+}
+
+impl<T: MeshField> DefaultEncoding for Receiver<T> {
+    type Encoding = PortField;
+}
+
+impl<T: MeshField> From<Port> for Receiver<T> {
+    fn from(port: Port) -> Self {
+        Self::from_port(port)
+    }
+}
+
+impl<T: MeshField> From<Receiver<T>> for Port {
+    fn from(receiver: Receiver<T>) -> Self {
+        receiver.into_port()
+    }
+}
+
+impl<T: MeshField> Receiver<T> {
+    /// Bridges this and `sender` together, consuming both `self` and `sender`.
+    ///
+    /// See [`Sender::bridge`] for more details.
+    pub fn bridge(self, sender: Sender<T>) {
+        sender.bridge(self)
+    }
+
+    fn into_port(self) -> Port {
+        // SAFETY: the encode function is for `T`.
+        unsafe { self.0.into_port(Sender::<T>::encode_message) }
+    }
+
+    fn from_port(port: Port) -> Self {
+        Self(
+            // SAFETY: the vtable and new handler function are for a queue with
+            // type `T`.
+            unsafe {
+                ReceiverCore::from_port(
+                    port,
+                    const { &ElementVtable::new::<T>() },
+                    RemotePortHandler::new::<T>,
+                )
+            },
+            PhantomData,
+        )
+    }
```

Comment:
I think I know why this loop is required, but some clarification as to the specific condition this is guarding against seems useful

Reply:
The comment doesn't line up with the code. Is this the loop in access?

Reply:
_sigh_ thanks github. yes - this was supposed to be in access